httputil/httputil.go

```diff
@@ -18,9 +20,32 @@ var (
 	DefaultTransport = http.DefaultTransport
 	UserAgent = "Bazelisk"
 	linkPattern = regexp.MustCompile(`<(.*?)>; rel="(\w+)"`)
+
+	RetryClock = Clock(&realClock{})
+	MaxRetries = 4
+	// MaxRequestDuration defines the maximum amount of time that a request and its retries may take in total
+	MaxRequestDuration = time.Second * 10
+	retryHeaders = []string{"Retry-After", "X-RateLimit-Reset", "Rate-Limit-Reset"}
 )
 
+type Clock interface {
+	Sleep(time.Duration)
+	Now() time.Time
+}
+
+type realClock struct {}
+
+func (*realClock) Sleep(d time.Duration) {
+	time.Sleep(d)
+}
+
+func (*realClock) Now() time.Time {
+	return time.Now()
+}
+
 // ReadRemoteFile returns the contents of the given file, using the supplied Authorization token, if set. It also returns the HTTP headers.
+// If the request fails with a transient error it will retry the request for at most MaxRetries times.
+// It obeys HTTP headers such as "Retry-After" when calculating the start time of the next attempt. If no such header is present, it uses an exponential backoff strategy.
```

Comment:
nit: line-break after "... of the next attempt."?

Reply:
Done